Seventh homework branch

Serg0420 opened the pull request.

Serg0420 committed:
seventh homework rep comm1
reworked UsersInfoRepositoryImpl and some clean architecture. Now using BuildConfig resources for some string constants instead context

Serg0420 committed:
seventh homework rep comm2
added map

Serg0420 committed:
seventh homework rep comm3
added List of users owned games

Serg0420 committed:
seventh homework rep comm4
rescaled all images

Serg0420 committed:
seventh homework rep comm5
added SettingFragment screen, in which you can enter your own Steam Id. Now all fragments are working whit saved in SharedPreferences Steam Id

Serg0420 committed:
seventh homework rep comm6
added service that can track users friend online status

Serg0420 committed:
seventh homework rep comm7
BIG UPDATE
now status tracker can be stopped in the app, and starts only on button clicking.
Added Room database to store users friends locations, that are showing on the map now
fixed some annoying bugs

Serg0420 committed:
seventh homework rep comm8
added bottom sheet on map, that will appear after click on geotag

Serg0420 committed:
seventh homework rep comm9
now in bottom sheet you can delete tag

vgoloveshchenko commented on app/build.gradle:
```diff
@@ -25,6 +28,21 @@ android {
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+            buildConfigField "String", "STEAM_API_BASE_URL","\"http://api.steampowered.com/\""
+            buildConfigField "String", "RELATIONSHIP","\"friend\""
+            buildConfigField "String", "MY_STEAM_ID","\"76561198277362398\""
+            buildConfigField "String", "STEAM_API_KEY","\"EA8CD62FA7C21007003DEF64FF96E20C\""
+            buildConfigField "String", "STEAM_API_IMG_LOGO_URL","\"http://media.steampowered.com/steamcommunity/public/images/apps/\""
```
если configfiled одинаковый для разных типов, то можно положить в блоке defaultConfig, константы все равно появятся

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/data/repository/GamesRepositoryImpl.kt:
```diff
@@ -0,0 +1,20 @@
+        return steamApi.getUserGames(
+            BuildConfig.STEAM_API_KEY,
+            steamId,
+            "true"
```
в константы, чтобы было понятно что это флаг делает, либо для предоставления гибкости можно сделать опциональный параметр в апи, тогда будет так 
```
@GET("IPlayerService/GetOwnedGames/v0001/")
    suspend fun getUserGames(
        @Query("key") key: String,
        @Query("steamid") steamid: String,
        @Query("include_appinfo") includeAppInfo: Boolean = true
    ): GamesResponse
```

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/data/api/SteamApi.kt:
```diff
@@ -19,4 +20,11 @@ interface SteamApi {
+    suspend fun getUserGames(
+        @Query("key") key: String,
+        @Query("steamid") steamid: String,
+        @Query("include_appinfo") includeAppInfo: String
```
все параметры будут подставляться как есть, то есть ты здесь можешь использовать и boolean и long, не обязательно только строки

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/data/mapper/GameMapper.kt:
```diff
@@ -0,0 +1,17 @@
+            appid = it.appid,
+            name = it.name,
+            playtimeForever = it.playtimeForever,
+            imgIconUrl = BuildConfig.STEAM_API_IMG_LOGO_URL+it.appid+"/"+it.imgIconUrl+".jpg",
```
можно сделать по-другому, STEAM_API_IMG_LOGO_URL это будет не кусок строки, а паттерн, в который ты будешь добавлять части
что-то вроде такого `https://steam.api/%s/%s.jpg`
тогда для получения значения `STEAM_API_IMG_LOGO_URL.format(it.appid, it.imgIconUrl)`

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/details/DetailsFragment.kt:
```diff
@@ -0,0 +1,120 @@
+    private val binding
+        get() = requireNotNull(_binding) { "View was destroyed" }
+
+    private val viewModel by inject<DetailsViewModel>()
```
```suggestion
    private val viewModel by viewModel<DetailsViewModel>()
```

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/details/DetailsFragment.kt:
```diff
@@ -0,0 +1,120 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        val args: DetailsFragmentArgs by navArgs()
```
аргументы можно положить в поле фрагмента, они все равно будут через делегат доставаться лениво

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/details/DetailsFragment.kt:
```diff
@@ -0,0 +1,120 @@
+
+            statusTxtv.append(status)
+
+            toolbar.setNavigationOnClickListener {
+                findNavController().navigateUp()
+            }
```
можно проще 
```suggestion
            toolbar.setupWithNavController(findNavController())
```

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/details/DetailsFragment.kt:
```diff
@@ -0,0 +1,120 @@
+                findNavController().navigateUp()
+            }
+
+            viewModel
+                .dataFlow
+                .flowWithLifecycle(viewLifecycleOwner.lifecycle, Lifecycle.State.STARTED)
+                .launchIn(viewLifecycleOwner.lifecycleScope)
```
а что тут будет? ты подписываешься и с данными ничего не делаешь

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/details/DetailsViewModel.kt:
```diff
@@ -0,0 +1,23 @@
+        replay = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+
+    val dataFlow = locationFlow.onEach {
+        dataSource.insertUsers(it)
+    }
```
чтобы не делать подписку в фрагменте стоит сделать так
```
init {
    locationFlow
        .onEach(dataSource::insertUsers)
        .launchIn(viewModelScope)
}
```
тогда и флоу публичный не нужен

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/details/StateTrackerService.kt:
```diff
@@ -0,0 +1,154 @@
+        return runCatching {
+            retroDataSource.getUserInfoById(userSteamId)
+        }
+            .fold(
+                onSuccess = { LCE.Content(it) },
+                onFailure = { LCE.Error(it) }
+            )
```
в сервисе ты я так понимаю можешь работать прямо с Result, Lce здесь лишний, у тебя прогресса никакого нет, если бы ты допустим обновлял нотификацию когда что-то грузится, то ок, а так можно убрать

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/details/StateTrackerService.kt:
```diff
@@ -0,0 +1,154 @@
+    private fun startStatusTracking() {
+        statusJob = coroutineScope.launch {
+            while (true) {
+                when (val lce = runCatch()) {
+                    is LCE.Error -> {
+                        handleError(lce.throwable.toString())
+                    }
+                    is LCE.Content -> {
+                        currentOnlineStatus = lce.data.getStatus(this@StateTrackerService)
+                    }
+                }
```
если сделать по комментарию ниже, то тут будет так
```
request()
    .onSuccess { //logic }
    .onFailure { //logic }
```

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/friendslst/FriendsFragment.kt:
```diff
@@ -38,12 +40,13 @@ class FriendsFragment : Fragment() {
-        get() = requireNotNull(_binding) {
-            "View was destroyed"
-        }
+    private val viewModel by inject<FriendsViewModel> {
```
```suggestion
    private val viewModel by viewModel<FriendsViewModel> {
```

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/ownedgames/GameAdapter.kt:
```diff
@@ -0,0 +1,45 @@
+
+class GameAdapter(
+    context: Context
+) : ListAdapter<Game, RecyclerView.ViewHolder>(DIFF_UTIL) {
```
в адаптере используется 1 тип вью холдера, не зря тут параметризация есть, можно так, чтобы проверок лишних не писать
```suggestion
) : ListAdapter<Game, GameViewHolder>(DIFF_UTIL) {
```

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/ownedgames/GameAdapter.kt:
```diff
@@ -0,0 +1,45 @@
+        val item = getItem(position)
+
+        if (holder is GameViewHolder) {
+            if (item !is Game) return
```
параметризация элементом Game, разве может другой тип придти?

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/ownedgames/GamesFragment.kt:
```diff
@@ -0,0 +1,110 @@
+        requireContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
+    }
+
+    private val viewModel by inject<GamesViewModel>{
```
```suggestion
    private val viewModel by viewModel<GamesViewModel>{
```

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/ownedgames/GamesFragment.kt:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    private fun refresh() {
+        adapter.submitList(emptyList())
+        binding.tryAgainBtn.isVisible = false
+        binding.progressIndicator.isVisible = false
```
этим должна управлять вью модель через подписку фрагмента, фрагмент сам никакую логику не делает

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/profile/MyProfileFragment.kt:
```diff
@@ -10,27 +11,33 @@ import androidx.lifecycle.Lifecycle
+        requireContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
+    }
+
+    private val viewModel by inject<MyProfileViewModel> {
```
```suggestion
    private val viewModel by viewModel<MyProfileViewModel> {
```

vgoloveshchenko commented on app/src/main/java/com/example/steamapp/presentation/ui/profile/MyProfileFragment.kt:
```diff
@@ -10,27 +11,33 @@ import androidx.lifecycle.Lifecycle
         get() = requireNotNull(_binding) { "View was destroyed" }
 
-    private val viewModel by inject<MyProfileViewModel>()
+    private val sharedPreferences by lazy {
+        requireContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
+    }
```
уже было место использования shared preferences, стоит сделать 1 сервис и положить туда все настройки, и потом по интерфейсу или интерфейсам инжектить

Serg0420 committed:
seventh homework rep comm10
fixed some clean issues, shared preferences now have there own repository, so current user id is injecting in fragments by its interface now.
